scipy/optimize/_optimize.py

```diff
@@ -3477,12 +3489,23 @@ def brute(func, ranges, args=(), Ns=20, full_output=0, finish=fmin,
     for k in range(N):
         if type(lrange[k]) is not type(slice(None)):
             if len(lrange[k]) < 3:
-                lrange[k] = tuple(lrange[k]) + (complex(Ns),)
+                if integrality[k]:
+                    ub = np.floor(lrange[k][1])
+                    lb = min(np.ceil(lrange[k][0]), ub)
+                    lrange[k] = lb, ub
+                    n_ints_between_bounds = lrange[k][1] - lrange[k][0] + 1
```

Comment:
This code block is going to end up evaluating points outside the `(low, high)` limits.
e.g. 
```
lrange = [(0.3, 0.9)]
k = 0
Ns = 20

ub = np.floor(lrange[k][1])
lb = min(np.ceil(lrange[k][0]), ub)
lrange[k] = lb, ub
n_ints_between_bounds = lrange[k][1] - lrange[k][0] + 1
n_points = min(Ns, n_ints_between_bounds)

print(lrange, n_points, n_ints_between_bounds)
```
gives:
```([(0.0, 0.0)], 1.0, 1.0)```

In this example both `n_points` and `n_ints_between_bounds` should be zero, with an Exception being raised because there are no ints. It shouldn't be evaluating a point at 0.

If you have `lb = np.ceil(lrange[k][0])` then `n_points` would come out as zero, which you could check for and raise an Exception. If you do decide to follow this, then you may need to check that low < high to begin with:
```
# place at top of if-clause
lrange[k][0], lrange[k][1] = min(lrange[k][0], lrange[k][1]), max(lrange[k][0], lrange[k][1])
# or this might work
lrange[k] = np.sort(lrange[k])
```

Reply:
Yup, good catch. I'll take a look.

Reply:
Fixed this by converting `lrange[k]` into a 1D grid (`grid`) then doing the most obvious thing:
```
    temp = np.unique(np.round(grid))
    if temp[0] < grid[0]:
        temp = temp[1:]
    if temp[-1] > grid[-1]:
        temp = temp[:-1]
    return temp
```
I wanted to avoid having to use `unique`, as it doesn't make much sense to create an array full of values that won't be used, but it makes the logic pretty obvious.

This does the right thing as long as `grid` isn't empty to begin with.